File: python/aurigraph_sdk/namespaces/governance.py
"""Namespace for /api/v11/governance endpoints -- on-chain governance."""

from __future__ import annotations

from typing import TYPE_CHECKING, Any
from urllib.parse import quote

from aurigraph_sdk.models import Proposal, TreasuryStats, VoteReceipt

if TYPE_CHECKING:
    from aurigraph_sdk.client import AurigraphClient


class GovernanceApi:
    """On-chain governance operations."""

    def __init__(self, client: AurigraphClient) -> None:
        self._client = client

    def list_proposals(self) -> list[Proposal]:
        """List all governance proposals."""
        data = self._client._get("/governance/proposals")
        items: list[dict[str, Any]] = data if isinstance(data, list) else data.get("proposals", [])  # type: ignore[assignment]
        return [Proposal.model_validate(p) for p in items]

    def get_proposal(self, proposal_id: str) -> Proposal:
        """Get a specific governance proposal.

        Args:
            proposal_id: The proposal identifier.
        """
        enc = quote(proposal_id, safe="")
        data = self._client._get(f"/governance/proposals/{enc}")
        return Proposal.model_validate(data)

    def vote(self, proposal_id: str, approve: bool) -> VoteReceipt:
        """Cast a vote on a governance proposal.

        Args:
            proposal_id: The proposal to vote on.
            approve: True to vote in favor, False to vote against.

        Returns:
            Vote receipt confirming the vote was recorded.
        """
        enc = quote(proposal_id, safe="")
        data = self._client._post(
            f"/governance/proposals/{enc}/vote", {"approve": approve}
        )
        return VoteReceipt.model_validate(data)

    def get_treasury(self) -> TreasuryStats:
        """Get governance treasury statistics."""
        data = self._client._get("/governance/treasury")
        return TreasuryStats.model_validate(data)

File: rust/src/namespaces/governance.rs
//! On-chain governance namespace for `/api/v11/governance/*` endpoints.
//!
//! Provides proposal listing, voting, and treasury inspection.
//!
//! # Example
//!
//! ```rust,no_run
//! # async fn example(client: &aurigraph_sdk::AurigraphClient) -> aurigraph_sdk::Result<()> {
//! let proposals = client.governance().list_proposals().await?;
//! println!("Active proposals: {}", proposals);
//!
//! let receipt = client.governance().vote("prop-1", "FOR").await?;
//! println!("Vote receipt: {}", receipt);
//! # Ok(())
//! # }
//! ```

use crate::client::AurigraphClient;
use crate::error::Result;

/// Governance API namespace.
///
/// Obtained via [`AurigraphClient::governance()`].
#[derive(Debug)]
pub struct GovernanceApi<'a> {
    client: &'a AurigraphClient,
}

impl<'a> GovernanceApi<'a> {
    pub(crate) fn new(client: &'a AurigraphClient) -> Self {
        Self { client }
    }

    /// List all governance proposals.
    ///
    /// Calls `GET /api/v11/governance/proposals`.
    pub async fn list_proposals(&self) -> Result<serde_json::Value> {
        self.client.get("/governance/proposals").await
    }

    /// Get a single governance proposal by ID.
    ///
    /// Calls `GET /api/v11/governance/proposals/{proposal_id}`.
    pub async fn get_proposal(&self, proposal_id: &str) -> Result<serde_json::Value> {
        self.client
            .get(&format!("/governance/proposals/{}", proposal_id))
            .await
    }

    /// Cast a vote on a governance proposal.
    ///
    /// Calls `POST /api/v11/governance/proposals/{proposal_id}/vote`.
    pub async fn vote(&self, proposal_id: &str, vote: &str) -> Result<serde_json::Value> {
        let body = serde_json::json!({ "vote": vote });
        self.client
            .post(
                &format!("/governance/proposals/{}/vote", proposal_id),
                &body,
            )
            .await
    }

    /// Get governance treasury statistics.
    ///
    /// Calls `GET /api/v11/governance/treasury`.
    pub async fn get_treasury(&self) -> Result<serde_json::Value> {
        self.client.get("/governance/treasury").await
    }
}
